### PR TITLE
Add multipart zip support

### PR DESCRIPTION
**Pull request by iranroman**
This PR implements handling of multi-part zip files, which was the zip format of the `fsd50k` dataset and resulted in #55 . This will be needed to write the loaders for `TAU NIGENS SSE 2021`, `TAU NIGENS SSE 2020`, and `TAU SSE 2019`.

### CODE REVIEW

File: soundata/download_utils.py [hunk 1]
```diff
@@ -94,36 +97,49 @@ def downloader(
...
+
+                        move_directory_contents(source_dir, destination_dir)
+
+            elif isinstance(remotes[k], list):
+
+                for l in range(len(remotes[k])):
+                    logging.info("[{}] downloading {}".format(k, remotes[k][l].filename))
+                    download_from_remote(remotes[k][l], save_dir, force_overwrite)
+                zip_path = save_dir+"/"+k+".zip"
+                outpath = save_dir+"/"+k+"_single.zip"
```
Comment (genisplaja) [inline soundata/download_utils.py]: Maybe using `os.path.join()` here?

**Comment by genisplaja** [comment]
Thanks @iranroman for that!:) 

Regarding your last comment on #55: yes of course it has no sense for the user to get just some data splits instead of downloading them all. In my version of the FSD50K loader, I added `Exceptions` to make sure that users were not allowed to download just a single split alone but the entire `development` or `evaluation` splits. However, I like the idea of storing the `RemoteFileMetadata` of the splits as a `list` to identify the multipart zips, and then using the expected final folder name as key.

However, there is something I don't understand: where are we testing that merging and unzipping code now? Should this be done in `test_download_utils.py`?

**Comment by iranroman** [comment]
Hello @genisplaja. Thanks for your feedback.

> However, there is something I don't understand: where are we testing that merging and unzipping code now? Should this be done in `test_download_utils.py`?

Isn't it implied that the download, merging, and unzipping all were done correctly when the `pytest -s tests/test_full_dataset.py --local --dataset fsd50` passes?

File: soundata/download_utils.py (hunk 1 above)
Comment (iranroman) [reply soundata/download_utils.py]: Fixed!

**Comment by iranroman** [comment]
Hello @genisplaja. I followed your suggestion and I have added a multipart zip test in `test_download_utils.py`. I also refactored the code to create a `download_multipart_zip` function in `download_utils.py`. Let me know if you have more suggestions.

**Comment by genisplaja** [comment]
Looks good @iranroman!! Thanks for that:) You need to make sure that every line is covered by tests, even if the `test_full_dataset` passes in local! Codecov is checking that for you every time you update branch in GitHub. Any idea why the circleCI tests are not passing here? I cloned your repo and all tests all passing locally for me. Have you done black formatting?

**Comment by iranroman** [comment]
No idea @genisplaja. This is my first PR int his repo. I was expecting them to just pass. You would know better than me.

**Comment by iranroman** [comment]
Hello all, `codecov` is complaining about lines that I did not originally write. What should I do to make `codecov` ignore these lines as before?

**Comment by magdalenafuentes** [comment]
@iranroman will look into this tomorrow!

File: soundata/download_utils.py [hunk 2]
```diff
@@ -54,6 +55,8 @@ def downloader(
             The directory to download the data
         remotes (dict or None):
             A dictionary of RemoteFileMetadata tuples of data in zip format.
+            If a dictionary key has a list of RemoteFileMetadata,
```
Comment (magdalenafuentes) [inline soundata/download_utils.py]: nit: Maybe rephrase? `If an element of the dictionary is a list of RemoteFileMetadata`?

**Comment by iranroman** [comment]
Hello @genisplaja.

This PR is very relevant to the dataset loader you wrote for `fsd50k`. Would you like to have a look before we close it?

File: soundata/download_utils.py [hunk 3]
```diff
@@ -94,36 +97,44 @@ def downloader(
...
-            if ".zip" in extension:
-                download_zip_file(remotes[k], save_dir, force_overwrite, cleanup)
-            elif ".gz" in extension or ".tar" in extension or ".bz2" in extension:
-                download_tar_file(remotes[k], save_dir, force_overwrite, cleanup)
-            else:
-                download_from_remote(remotes[k], save_dir, force_overwrite)
 
-            if remotes[k].unpack_directories:
-                for src_dir in remotes[k].unpack_directories:
+            if isinstance(remotes[k], list):
```
Comment (magdalenafuentes) [inline soundata/download_utils.py:101]: Since you're assuming that the multiple files are `zip`, maybe you can check that and add an `NotImplementedError` if it is not?

**Comment by genisplaja** [comment]
Yes! I'll do that tomorrow morning (it's almost midnight here!). Thanks for letting me know:)

File: soundata/download_utils.py (hunk 3 above)
Comment (iranroman) [reply soundata/download_utils.py:101]: Thanks for this idea. I've implemented it

File: soundata/download_utils.py (hunk 2 above)
Comment (iranroman) [reply soundata/download_utils.py]: makes sense. Done

**Comment by iranroman** [comment]
No problem @genisplaja.

In the meantime I ran `pytest -s tests/test_full_dataset.py --local --dataset fsd50k` and everything passed for me (including download, multipart joining, and unzipping). 

The multipart zip download, joining, and unpacking now is handled by `download_utils.py`. Let us know your thoughts.

File: soundata/datasets/fsd50k.py [hunk 4]
```diff
@@ -206,51 +206,54 @@
     primaryClass={cs.SD}
 }
 """
+
+# a dictionary key that hasa a list of RemoteFileMetadata implies a multi-part zip
```
Comment (genisplaja) [inline soundata/datasets/fsd50k.py]: typo: has a

File: soundata/download_utils.py [hunk 5]
```diff
@@ -140,6 +154,37 @@ def update_to(self, b=1, bsize=1, tsize=None):
...
+            Path to save downloaded file
+        force_overwrite (bool):
+            If True, overwrites existing files
+        cleanup (bool):
+            If True, remove zipfile after unziping
+
+    """
+    for l in range(len(zip_remotes)):
+        download_from_remote(zip_remotes[l], save_dir, force_overwrite)
+    zip_path = os.path.join(save_dir, obj_to_download + ".zip")
```
Comment (genisplaja) [inline soundata/download_utils.py]: This is a good way to get the principal zip filename to use it as input of command `zip`, but it forces the use of the principal filename without the `.zip` extension as key of the list of parts of the split compressed file in the `REMOTES` dict (see line 213-214 of FSD50K loader). Am I right?
If so, this may cause confusion for an external contributor, since in other loaders one would just put, for instance, `development` as key, as I did for the original FSD50K loader. This should be specified somewhere in the docs, or I would suggest to iterate over the download item list parsing the field `filename` to identify the one with `.zip` extension, which is actually the input filename for the `zip` command. I think it would be something like:
```
zip_path = next((part.filename for part in zip_remotes if '.zip' in part.filename), None)
```
and then you can get the output path like:
```
out_path = zip_path.replace('.zip', '_single.zip')
```
Then the input `obj_to_download` would be no longer needed. Since at the moment this function is just for multipart files in zip format, it should work! And then the contributor is allowed to use any key for the divided file.
Also because I think that you don't actually check that the actual filename is `k + '.zip'`,  or `obj_to_download + '.zip'`.
Just a proposal:) What do you think?

Comment (iranroman) [reply soundata/download_utils.py]: Hello, thanks for your suggestions. 

I'm afraid I tried making the changes you suggested and now the tests are not passing. do you wanna give it a shot and open a PR against this one before we join with `main`?

Comment (iranroman) [reply soundata/download_utils.py]: Never mind. I got it. Great suggestions, by the way. Thanks a lot!

Comment (genisplaja) [reply soundata/download_utils.py]: Great, no worries! This PR looks ready to me.

**Review by magdalenafuentes — APPROVED** [review]
